File: src/material/metal.rs
use super::super::{
    utilities::prelude::*,
    ray::Ray,
    hittable::prelude::*,
    material::prelude::*,
};

#[derive(Clone)]
pub struct Metal {
    pub albedo : RGB,
    pub fuzz   : f64,
}

impl Metal {
    pub fn new(albedo: RGB, fuzz: f64) -> Metal {
        Self { albedo, fuzz }
    }
}

impl Material for Metal {
    fn scatter(&self, ray: &Ray, rec: &HitRecord, rng: &mut SmallRng) -> Option<ScatterRecord> {
        let reflected = ray.direction.unit().reflect_on(rec.normal) + self.fuzz * Vec3::random_in_unit_sphere(rng);
        Some(ScatterRecord::Specular {
            specular_ray: Ray::new(rec.p, reflected, 0.0),
            attenuation: self.albedo,
        })
    }
}

File: src/material/metal.cpp
#include "metal.hpp"

auto Metal::scatter(const Ray &ray, const HitRecord &rec) const -> std::optional<std::tuple<Ray, RGB, f64>> {
    Vec3 reflected = ray.direction.unit().reflect_on(rec.normal) + fuzz * Vec3::random_in_unit_sphere();
    if (Vec3::dot(reflected, rec.normal) > 0.0)
        return std::optional(std::make_tuple(Ray(rec.p, reflected), albedo, 0.0));
    return std::optional<std::tuple<Ray, RGB, f64>>();
}
